### ISSUE TITLE
Binary Data Endpoints

### ISSUE THREAD

**Issue by sirinath** [edgedb/edgedb#1305]
For low latency, high throughput application can you provide Binary Data Endpoints using standards like:

- [Fast Binary Encoding (FBE)](https://chronoxor.github.io/FastBinaryEncoding/)
- [Simple Binary Encoding (SBE)](https://real-logic.github.io/simple-binary-encoding/)
- [gRPC](https://grpc.io/)
- [MessagePack](https://msgpack.org/)
- [Concise Binary Object Representation (CBOR)](http://cbor.io/)

There many client libraries in different languages for the above hence you do not need drivers to connect as added benefits.

**Comment by 1st1** [comment]
Yeah, we're considering this, it's on our roadmap for post 1.0 releases.

**Comment by LeonardAukea** [comment]
+1 for gRPC

**Comment by alelazar** [comment]
Any updates on this?

**Comment by 1st1** [comment]
It is indeed unclear. Since this issue had been opened our protocol has evolved a lot. It now implements automatic connection recovery, automatic transaction retries, and many other things. It's been optimized to keep the number of roundtrips to the minimum (usually it's just 1) and the data unmarshalling in our clients is as efficient as possible.

The only benefit of implementing something like gRPC natively would be avoiding re-serialization of the data on the client side, but we'd have to implement serialization in the DB itself and that will add some overhead (not to mention I doubt we can do that without patching Postgres). Then we need to find a way to describe EdgeDB's schema in protocol buffer format (or reflect to it).

All of this is certainly doable, but it's a lot of effort. Currently we don't have bandwidth to implement this, but that might change around EdgeDB 4.0 or so.

**Comment by alelazar** [comment]
Thanks for the update!

All of my questions here are just to clarify things that I'm either misunderstanding or not finding in docs. 

When you say "without patching Postgres" is the binary protocol something from edgedb or from postgres? 

As far as efficiency goes, that existing binary protocol is approximately on par with gRPC? When you say "avoiding re-serialization of the data on the client side," you mean that if I query the DB on my server, then send that data onwards somewhere else, my server just did the work to deserialize "binary protocol" then serialize it again in gRPC/REST whatever? So it's slightly more work for something that sits in between. But the db-server hop is efficient. 

While the [HTTP protocol](https://www.edgedb.com/docs/clients/http/protocol) exists, this is just for languages that don't have a client library? The published client libraries all handle the binary protocol for us?

Thanks again!
